lib/HLSL/HLOperations.cpp

```diff
@@ -530,6 +558,79 @@ static void SetHLFunctionAttribute(Function *F, HLOpcodeGroup group,
     // No default attributes for these opcodes.
     break;
   }
+  assert(!(attribs.hasAttribute(AttributeSet::FunctionIndex,
+                                Attribute::ReadNone) &&
+           attribs.hasAttribute(AttributeSet::FunctionIndex,
+                                Attribute::ReadOnly)) &&
+         "conflicting ReadNone and ReadOnly attributes");
+  return attribs;
+}
+
+static std::string GetHLFunctionAttributeMangling(const AttributeSet &attribs) {
+  std::string mangledName;
+  raw_string_ostream mangledNameStr(mangledName);
+
+  // Capture for adding in canonical order later.
+  bool ReadNone = false;
+  bool ReadOnly = false;
+  bool ArgMemOnly = false;
+  bool NoDuplicate = false;
+  bool WaveSensitive = false;
+
+  // Ensure every function attribute is recognized.
+  for (unsigned Slot = 0; Slot < attribs.getNumSlots(); Slot++) {
+    if (attribs.getSlotIndex(Slot) == AttributeSet::FunctionIndex) {
+      for (auto it = attribs.begin(Slot), e = attribs.end(Slot); it != e;
+           it++) {
+        if (it->isEnumAttribute()) {
+          switch (it->getKindAsEnum()) {
+          case Attribute::ReadNone:
+            ReadNone = true;
+            break;
+          case Attribute::ReadOnly:
+            ReadOnly = true;
+            break;
+          case Attribute::ArgMemOnly:
+            ArgMemOnly = true;
+            break;
+          case Attribute::NoDuplicate:
+            NoDuplicate = true;
+            break;
+          case Attribute::NoUnwind:
+            // All intrinsics have this attribute, so mangling is unaffected.
+            break;
+          default:
+            assert(false && "unexpected attribute for HLOperation");
+          }
+        } else if (it->isStringAttribute()) {
+          StringRef Kind = it->getKindAsString();
+          if (Kind == HLWaveSensitive) {
+            assert(it->getValueAsString() == "y" &&
+                   "otherwise, unexpected value for WaveSensitive attribute");
+            WaveSensitive = true;
+          } else {
+            assert(Kind != "ExtractRecordStructFromArray" &&
+                   "unexpected string function attribute for HLOperation");
+          }
+        }
+      }
+    }
+  }
+
+  // Validate attribute combinations.
+  assert(!(ReadNone && ReadOnly && ArgMemOnly) &&
+         "ReadNone, ReadOnly, and ArgMemOnly are mutually exclusive");
+
+  // Add mangling in canonical order
+  if (NoDuplicate)
+    mangledNameStr << "nd";
+  if (ReadNone)
+    mangledNameStr << "rn";
+  if (ReadOnly)
+    mangledNameStr << "ro";
```

Comment:
Did you want to add ArgMemOnly "amo" as part of this merge, or separately?  If separately, is there an issue tracking that?